`pallet-message-queue`: add queue pausing

ggwpez opened the pull request:
Introduces a `QueuePausedQuery` hook into the MQ pallet that is polled once before starting to service a queue. This hook decided whether the queue should start service or not. 

It can be used to dynamically pause queues and is better than using `Yield` since no PoV is wasted for loading the first message and `ProcessMessage` does not need to be implemented just for pausing.  
Context: Greatly reduces the complexity of https://github.com/paritytech/cumulus/pull/2157/

I also implemented explicit suspension (via `set_suspended`) by adding a bool flag to the `BookState` [here](https://github.com/paritytech/substrate/compare/oty-mq-queue-suspension?expand=1). This is generally a nice thing to have for the future IMHO, but not well usable for 2157. This approach works better when knowing in advance which queues need to stay paused without any more complicated predicates involved. 
2157 instead has custom predicate logic [here](https://github.com/paritytech/cumulus/blob/741433a9f8d475fd140b9ddb543d14a4814228ea/pallets/xcmp-queue/src/lib.rs#L630) which can change at any time.

**Changes:**
- :rotating_light: Add `QueuePausedQuery` to the MQ pallet Config. `()` indicates that nothing is paused.
- Add trait `frame_support::traits::QueuePausedQuery`
- Extend `pallet_message_queue::Error` and `ExecuteOverweightError`

Polkadot companion: https://github.com/paritytech/polkadot/pull/7433

ggwpez committed:
pallet-message-queue: add queue pausing

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2954933

ggwpez committed:
Fix build

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

gavofyork reviewed (approved):
Seem harmless enough, but what about migrations?

ggwpez commented:
> Seem harmless enough, but what about migrations?

Its not changing any storage - just adding a `QueuePausedQuery` hook.

kianenigma commented on frame/support/src/traits/messages.rs:
```diff
@@ -220,3 +230,15 @@ where
 }
+
+/// Provides information on paused queues.
+pub trait QueuePausedQuery<Origin> {
```
As a design nit, I sometimes wonder if we should avoid creating a new trait like this when it can be easily expressed by `Convert<Origin, bool>` (cc @gpestana this is similar to my comment in https://github.com/paritytech/substrate/pull/13983#discussion_r1187519009 🙈).

No strong opinion myself, but I might personally lean into reusing `Convert` more and more.

ggwpez replied:
Interesting idea. Yes I also dont like having tons of trivial traits, just for naming.  

The only problem i see with `Convert` is on the implementor side. When seeing `impl<T: Config> Convert… for Pallet<T>`, it is not clear what it means. Maybe we can use type alias for that, or something similar.

gpestana replied:
Maybe we could use the `frame_support::Get` trait instead? The type name is explicitly enough to use the getter trait and it things readable.

ggwpez replied:
Suppose we have `impl Get<bool> for Pallet<T>`.  
Now from the implementor side what does this mean? Anyone seeing this code will have no idea what it means.

kianenigma reviewed (approved):
LGTM. I would suggest double checking that all code-paths leading into servicing queues is blocked, as I could not be sure that the places that are already checked are not missing anything.

ggwpez committed:
Remove check

Otherwise it would not start servicing queues that started paused
and became unpaused afterwards.

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

gupnik commented on frame/message-queue/src/lib.rs:
```diff
@@ -473,6 +473,13 @@ pub mod pallet {
+		/// This also applies to manual servicing via `execute_overweight` and `service_queues`. The
+		/// value of this is only polled once before servicing the queue. This means that changes to
+		/// it that happen *within* the servicing will not be reflected.
+		type QueuePausedQuery: QueuePausedQuery<<Self::MessageProcessor as ProcessMessage>::Origin>;
```
Can we enforce an error if someone tries to change this within the servicing?

ggwpez replied:
Sure we could check this predicate every time before continuing service, but that is what a `Yield` return can already be used for.

gpestana commented on frame/support/src/traits/messages.rs:
```diff
@@ -220,3 +230,15 @@ where
+/// Provides information on paused queues.
+pub trait QueuePausedQuery<Origin> {
+	/// Whether this queue is paused.
+	fn is_paused(origin: &Origin) -> bool;
```
In which cases would the origin be useful as an input here?

ggwpez replied:
The `origin` is actually the id of the queue. The two terms are used interchangeably throughout the pallet.

gpestana reviewed (approved):
Added a question and a suggestion wrt to the `QueuePausedQuery` trait, but lgtm!

ggwpez commented:
bot merge
